File: GoTorch/backend/ndarray.py
"""Outwarding facing interface for tensor data representation

Defines the information relating to a tensor and stores its
data.

Attributes:
    strides: The stride shape for the tensor array.
    data: The contiguous array level data.
    shape: The shape of the tensor being represented.
    offset: The memory level offset.
"""


from __future__ import annotations

from typing import Sequence


class NDArray:
    """Multidimensional array backed by a flat contiguous list of floats."""

    def __init__(
        self,
        data: list[float] | Sequence[float],
        shape: tuple[int, ...],
        strides: tuple[int, ...] | None = None,
        offset: int = 0,
    ) -> None:
        self.data: list[float] = list(data)
        self.shape: tuple[int, ...] = tuple(shape)
        self.offset: int = offset

        if strides is None:
            self.strides: tuple[int, ...] = self._default_strides(self.shape)
        else:
            self.strides = tuple(strides)

    @staticmethod
    def _default_strides(shape: tuple[int, ...]) -> tuple[int, ...]:
        """Calculates standard C-contiguous (row-major) strides."""
        if not shape:
            return ()
        strides = [1] * len(shape)
        for i in range(len(shape) - 2, -1, -1):
            strides[i] = strides[i + 1] * shape[i + 1]
        return tuple(strides)

    def _index_to_offset(self, indices: tuple[int, ...]) -> int:
        """Maps multidimensional indices (i_0, i_1, ...) to flat buffer offset."""
        if len(indices) != len(self.shape):
            error_msg = (
               f"Index dimensional mismatch: expected {len(self.shape)} dims,"
               f" got {len(indices)}"
            )
            raise IndexError(error_msg)
        offset = self.offset
        for idx, stride in zip(indices, self.strides):
            offset += idx * stride
        return offset

    def __getitem__(self, indices: tuple[int, ...] | int) -> float:
        if isinstance(indices, int):
            indices = (indices,)
        return self.data[self._index_to_offset(indices)]

    def __setitem__(self, indices: tuple[int, ...] | int, value: float) -> None:
        if isinstance(indices, int):
            indices = (indices,)
        self.data[self._index_to_offset(indices)] = value

    def is_contiguous(self) -> bool:
        """Returns True if memory layout matches default row-major strides."""
        return self.strides == self._default_strides(self.shape) and self.offset == 0

    def to_contiguous(self) -> NDArray:
        """Returns a contiguous copy if strided/viewed, or self if already contiguous"""
        if self.is_contiguous():
            return self

        # Materialize non-contiguous views into standard order
        new_data: list[float] = []

        def _traverse(dim: int, current_idx: list[int]) -> None:
            if dim == len(self.shape):
                new_data.append(self[tuple(current_idx)])
                return
            for i in range(self.shape[dim]):
                current_idx.append(i)
                _traverse(dim + 1, current_idx)
                current_idx.pop()

        _traverse(0, [])
        return NDArray(new_data, self.shape)

    def transpose(self, dim0: int = 0, dim1: int = 1) -> NDArray:
        """Creates a zero-copy transposed view by swapping shape and strides."""
        if len(self.shape) < 2:
            return self

        new_shape = list(self.shape)
        new_strides = list(self.strides)

        new_shape[dim0], new_shape[dim1] = new_shape[dim1], new_shape[dim0]
        new_strides[dim0], new_strides[dim1] = new_strides[dim1], new_strides[dim0]

        # Shares the exact same self.data pointer/reference
        return NDArray(
            data=self.data,
            shape=tuple(new_shape),
            strides=tuple(new_strides),
            offset=self.offset,
        )

    def ones_like(self) -> NDArray:
        """Returns an array of ones with identical shape and standard strides."""
        size = 1
        for dim in self.shape:
            size *= dim
        return NDArray([1.0] * size, self.shape)

    def zeros_like(self) -> NDArray:
        """Returns an array of zeros with identical shape and standard strides."""
        size = 1
        for dim in self.shape:
            size *= dim
        return NDArray([0.0] * size, self.shape)

    def matmul(self, other: NDArray) -> NDArray:
        """Performs 2D matrix multiplication: (M, K) @ (K, N) -> (M, N)."""
        if len(self.shape) != 2 or len(other.shape) != 2:
            raise ValueError(
                    f"matmul requires 2D arrays, got {self.shape} and {other.shape}"
            )

        m, k1 = self.shape
        k2, n = other.shape

        if k1 != k2:
            raise ValueError(
                f"shape mismatch: inner dimensions {k1} and {k2} must match"
            )

        # Allocate flat memory for resulting (M, N) matrix
        out_data = [0.0] * (m * n)
        result = NDArray(data=out_data, shape=(m, n))

        # Dot product: C[i, j] = sum_k (A[i, k] * B[k, j])
        for i in range(m):
            for j in range(n):
                dot_sum = 0.0
                for k in range(k1):
                    dot_sum += self[i, k] * other[k, j]
                result[i, j] = dot_sum

        return result

    def add(self, other: NDArray) -> NDArray:
        """Element-wise addition with broadcasting (pad to the left with 1s)."""
        ndim_a = len(self.shape)
        ndim_b = len(other.shape)
        max_ndim = max(ndim_a, ndim_b)

        padded_a = (1,) * (max_ndim - ndim_a) + self.shape
        padded_b = (1,) * (max_ndim - ndim_b) + other.shape

        out_shape: list[int] = []
        for da, db in zip(padded_a, padded_b):
            if da == db:
                out_shape.append(da)
            elif da == 1:
                out_shape.append(db)
            elif db == 1:
                out_shape.append(da)
            else:
                raise ValueError(
                    f"shape mismatch: cannot broadcast dimensions {da} and {db}"
                )

        eff_strides_a: list[int] = [0] * max_ndim
        eff_strides_b: list[int] = [0] * max_ndim

        pad_a = max_ndim - ndim_a
        for i in range(pad_a, max_ndim):
            orig_d = i - pad_a
            if self.shape[orig_d] > 1:
                eff_strides_a[i] = self.strides[orig_d]

        pad_b = max_ndim - ndim_b
        for i in range(pad_b, max_ndim):
            orig_d = i - pad_b
            if other.shape[orig_d] > 1:
                eff_strides_b[i] = other.strides[orig_d]

        total_elements = 1
        for d in out_shape:
            total_elements *= d
        if not out_shape:
            total_elements = 0

        out_data: list[float] = [0.0] * total_elements
        if total_elements > 0:
            coord = [0] * max_ndim
            for out_idx in range(total_elements):
                off_a = self.offset
                off_b = other.offset
                for d in range(max_ndim):
                    off_a += coord[d] * eff_strides_a[d]
                    off_b += coord[d] * eff_strides_b[d]
                out_data[out_idx] = self.data[off_a] + other.data[off_b]

                if max_ndim > 0:
                    for d in range(max_ndim - 1, -1, -1):
                        coord[d] += 1
                        if coord[d] < out_shape[d]:
                            break
                        coord[d] = 0

        return NDArray(out_data, tuple(out_shape))

    def __add__(self, other: object) -> NDArray:
        if not isinstance(other, NDArray):
            return NotImplemented
        return self.add(other)

    def __radd__(self, other: object) -> NDArray:
        if not isinstance(other, NDArray):
            return NotImplemented
        return other.add(self)

File: GoTorch/backend/cppsrc/ndarray.cpp
#include "ndarray.hpp"

#include <algorithm>
#include <functional>
#include <stdexcept>

/**
 * @brief Computes standard row-major (C-contiguous) strides for a given shape.
 */
template <typename T>
std::vector<size_t> ndarray<T>::default_strides(const std::vector<size_t>& shape) {
    if (shape.empty()) {
        return {};
    }
    std::vector<size_t> str(shape.size(), 1);
    for (int i = static_cast<int>(shape.size()) - 2; i >= 0; --i) {
        str[i] = str[i + 1] * shape[i + 1];
    }
    return str;
}

/**
 * @brief Creates an independent deep copy of another ndarray and its storage.
 */
template <typename T>
void ndarray<T>::CreateDeepCopy(const ndarray& other) {
    shape = other.shape;
    strides = other.strides;
    offset = other.offset;

    if (other.storage) {
        storage = std::make_shared<Storage<T>>(other.storage->data);
    } else {
        storage = nullptr;
    }
}

/**
 * @brief Allocates and initializes ndarray storage from a raw buffer and shape.
 */
template <typename T>
void ndarray<T>::CreateNDArray(
    const T* array,
    const std::vector<size_t>& shape_dims,
    const std::vector<size_t>& custom_strides,
    size_t off
) {
    shape = shape_dims;
    offset = off;

    size_t total_size = num_elements(shape);

    if (custom_strides.empty()) {
        strides = default_strides(shape);
    } else {
        strides = custom_strides;
    }

    if (array != nullptr && total_size > 0) {
        storage = std::make_shared<Storage<T>>(array, total_size);
    } else {
        storage = std::make_shared<Storage<T>>(total_size);
    }
}

/**
 * @brief Constructs an ndarray by copying elements from a std::vector.
 */
template <typename T>
ndarray<T>::ndarray(
    const std::vector<T>& vec,
    const std::vector<size_t>& shape,
    const std::vector<size_t>& strides,
    size_t offset
) : offset(0) {
    this->shape = shape;
    this->offset = offset;
    this->strides = strides.empty() ? default_strides(shape) : strides;
    this->storage = std::make_shared<Storage<T>>(vec);
}

/**
 * @brief Creates a zero-copy transposed view by swapping two dimensions and their strides.
 */
template <typename T>
ndarray<T> ndarray<T>::transpose(size_t dim0, size_t dim1) const {
    if (shape.size() < 2) {
        return *this;
    }
    if (dim0 >= shape.size() || dim1 >= shape.size()) {
        throw std::out_of_range("Transpose dimension out of range");
    }
    std::vector<size_t> new_shape = shape;
    std::vector<size_t> new_strides = strides;
    std::swap(new_shape[dim0], new_shape[dim1]);
    std::swap(new_strides[dim0], new_strides[dim1]);
    return ndarray(storage, std::move(new_shape), std::move(new_strides), offset);
}

/**
 * @brief Maps multidimensional coordinate indices to a flat memory buffer offset.
 * 
 * 1. Validates coordinate count matches array rank
 * 2. Computes linear offset by summing base offset and coordinate-stride products
 * 
 * On coordinate count mismatch:
 * Throws std::out_of_range
 */
template <typename T>
size_t ndarray<T>::index_to_offset(const std::vector<size_t>& indices) const {
    if (indices.size() != shape.size()) {
        throw std::out_of_range(
            "Index dimensional mismatch: expected " +
            std::to_string(shape.size()) + " dims, got " +
            std::to_string(indices.size())
        );
    }
    size_t off = offset;
    for (size_t i = 0; i < indices.size(); ++i) {
        off += indices[i] * strides[i];
    }
    return off;
}

// Operation logic

namespace {

/**
 * @brief Advances a coordinate vector by one in row-major order.
 * 
 * 1. Iterates backwards from innermost dimension
 * 2. Increments coordinate; if within bounds, returns
 * 3. Resets coordinate to 0 and carries over to next outer dimension
 */
inline void advance_coordinate(std::vector<size_t>& coord, const std::vector<size_t>& shape) {
    for (int d = static_cast<int>(shape.size()) - 1; d >= 0; --d) {
        coord[d]++;
        if (coord[d] < shape[d]) {
            break;
        }
        coord[d] = 0;
    }
}

/**
 * @brief Pads a shape to the left with 1s to match the target rank.
 *
 * (3), target_rank=3 -> (1, 1, 3) 
 * No padding added if already target rank
 */
inline std::vector<size_t> pad_shape(const std::vector<size_t>& shape, size_t target_rank) {
    std::vector<size_t> padded(target_rank, 1);
    size_t pad = target_rank - shape.size();
    for (size_t i = 0; i < shape.size(); ++i) {
        padded[pad + i] = shape[i];
    }
    return padded;
}

/**
 * @brief Validates broadcast compatibility between two padded shapes and determines output shape.
 * 
 * 1. Allocates output shape vector of size max_ndim
 * 2. Compares corresponding dimensions of padded shapes
 * 3. Sets output dimension if equal or broadcasts if one dimension is 1
 * 4. Throws std::invalid_argument if dimensions are incompatible
 * 
 * On shape mismatch:
 * Throws std::invalid_argument
 */
inline std::vector<size_t> broadcast_shapes(
    const std::vector<size_t>& padded_a,
    const std::vector<size_t>& padded_b
) {
    size_t max_ndim = padded_a.size();
    std::vector<size_t> out_shape(max_ndim);

    // 2-3. Broadcastable if dimensions match or one of them is 1
    for (size_t i = 0; i < max_ndim; ++i) {
        size_t dim_a = padded_a[i];
        size_t dim_b = padded_b[i];
        if (dim_a == dim_b) {
            out_shape[i] = dim_a;
        } else if (dim_a == 1) {
            out_shape[i] = dim_b;
        } else if (dim_b == 1) {
            out_shape[i] = dim_a;
        } else {
            throw std::invalid_argument(
                "shape mismatch: cannot broadcast shapes"
            );
        }
    }
    return out_shape;
}

/**
 * @brief Computes effective broadcasted strides for an operand.
 * 
 * 1. Initializes stride vector of target_rank with 0s
 * 2. Computes left-padding offset from shape size and target_rank
 * 3. Maps non-broadcasted dimensions (size > 1) to original strides
 */
inline std::vector<size_t> compute_broadcast_strides(
    const std::vector<size_t>& shape,
    const std::vector<size_t>& strides,
    size_t target_rank
) {
    std::vector<size_t> eff_strides(target_rank, 0);
    size_t pad = target_rank - shape.size();

    // 3. Map non-broadcasted dimensions (size > 1) to original strides
    for (size_t i = 0; i < target_rank; ++i) {
        if (i >= pad) {
            size_t orig_dim = i - pad;
            if (shape[orig_dim] > 1) {
                eff_strides[i] = strides[orig_dim];
            }
        }
    }
    return eff_strides;
}

/**
 * @brief Allocates output buffer with total element count matching shape.
 */
template <typename T>
inline std::vector<T> allocate_output_buffer(const std::vector<size_t>& shape) {
    return std::vector<T>(ndarray<T>::num_elements(shape));
}

/**
 * @brief Accumulates element values from respective storage offsets with broadcasting.
 * 
 * 1. Verifies storage buffers for both operands are allocated
 * 2. Initializes coordinate tracking vector
 * 3. Computes buffer offsets using effective strides and stores element sums into output buffer
 * 4. Advances coordinates in row-major order
 * 
 * On unallocated storage:
 * Throws std::runtime_error
 */
template <typename T, typename BinaryOp>
inline void accumulate_broadcast(
    const std::shared_ptr<Storage<T>>& storage_a,
    size_t offset_a,
    const std::shared_ptr<Storage<T>>& storage_b,
    size_t offset_b,
    const std::vector<size_t>& eff_strides_a,
    const std::vector<size_t>& eff_strides_b,
    const std::vector<size_t>& out_shape,
    std::vector<T>& out_vec,
    BinaryOp op
) {
    size_t total_elements = out_vec.size();
    if (total_elements == 0) {
        return;
    }
    if (!storage_a || !storage_b) {
        throw std::runtime_error("NDArray storage is unallocated");
    }
    const T* a_ptr = storage_a->data.data() + offset_a;
    const T* b_ptr = storage_b->data.data() + offset_b;
    size_t max_ndim = out_shape.size();

    std::vector<size_t> coord(max_ndim, 0);
    for (size_t out_idx = 0; out_idx < total_elements; ++out_idx) {
        // 3. Compute strided offsets and apply binary operation
        size_t off_a = 0;
        size_t off_b = 0;
        for (size_t d = 0; d < max_ndim; ++d) {
            off_a += coord[d] * eff_strides_a[d];
            off_b += coord[d] * eff_strides_b[d];
        }
        out_vec[out_idx] = op(a_ptr[off_a], b_ptr[off_b]);

        advance_coordinate(coord, out_shape);
    }
}

/**
 * @brief Applies an element-wise binary operation across two ndarrays with broadcasting.
 * 
 * 1. Pads both input shapes to the left with 1s to match the maximum rank
 * 2. Validates broadcast dimension compatibility and determines output shape
 * 3. Computes effective strides for both operands (0 for broadcasted dimensions)
 * 4. Allocates output buffer and iterates through coordinates in row-major order
 * 5. Applies binary operation to element values from respective storage offsets
 * 
 * On shape mismatch:
 * Throws std::invalid_argument
 * On unallocated storage:
 * Throws std::runtime_error
 */
template <typename T, typename BinaryOp>
inline ndarray<T> broadcast_binary_op(
    const ndarray<T>& a,
    const ndarray<T>& b,
    BinaryOp op
) {
    size_t ndim_a = a.shape.size();
    size_t ndim_b = b.shape.size();
    size_t max_ndim = std::max(ndim_a, ndim_b);

    // 1. Pad shapes to the left with 1s so both have rank max_ndim
    std::vector<size_t> padded_a = pad_shape(a.shape, max_ndim);
    std::vector<size_t> padded_b = pad_shape(b.shape, max_ndim);

    // 2. Validate broadcast compatibility and determine output shape
    std::vector<size_t> out_shape = broadcast_shapes(padded_a, padded_b);

    // 3. Compute effective broadcasted strides for both operands
    std::vector<size_t> eff_strides_a = compute_broadcast_strides(a.shape, a.strides, max_ndim);
    std::vector<size_t> eff_strides_b = compute_broadcast_strides(b.shape, b.strides, max_ndim);

    // 4. Allocate output buffer
    std::vector<T> out_vec = allocate_output_buffer<T>(out_shape);

    // 5. Apply binary operation across strided coordinates
    accumulate_broadcast(a.storage, a.offset, b.storage, b.offset,
                         eff_strides_a, eff_strides_b, out_shape, out_vec, op);

    return ndarray<T>(out_vec, out_shape);
}

}  // namespace

/**
 * @brief Performs element-wise addition with broadcasting.
 */
template <typename T>
ndarray<T> ndarray<T>::add(const ndarray<T>& addend) const {
    return broadcast_binary_op(*this, addend, std::plus<T>{});
}

/**
 * @brief Performs element-wise subtraction with broadcasting.
 */
template <typename T>
ndarray<T> ndarray<T>::sub(const ndarray<T>& subtrahend) const {
    return broadcast_binary_op(*this, subtrahend, std::minus<T>{});
}

/**
 * @brief Performs 2D matrix multiplication over two ndarrays.
 * 
 * 1. Validates both operands are 2D arrays
 * 2. Checks inner dimensions match (K1 == K2)
 * 3. Allocates output buffer of size M x N
 * 4. Computes dot products across inner dimensions respecting strided layouts
 * 5. Returns resulting M x N ndarray
 * 
 * On non-2D input or inner dimension mismatch:
 * Throws std::invalid_argument
 */
template <typename T>
ndarray<T> ndarray<T>::matmul(const ndarray<T>& other) const {
    // 1. Validates both operands are 2D arrays
    if (shape.size() != 2 || other.shape.size() != 2) {
        throw std::invalid_argument(
            "matmul requires 2D arrays, got " +
            std::to_string(shape.size()) + " and " +
            std::to_string(other.shape.size())
        );
    }
    size_t m = shape[0];
    size_t k1 = shape[1];
    size_t k2 = other.shape[0];
    size_t n = other.shape[1];
    // 2. Checks inner dimensions match (K1 == K2)
    if (k1 != k2) {
        throw std::invalid_argument(
            "shape mismatch: inner dimensions " +
            std::to_string(k1) + " and " +
            std::to_string(k2) + " must match"
        );
    }
    // 3. Allocates output buffer of size M x N
    std::vector<T> out_data(m * n, 0);
    // 4. Computes dot products across inner dimensions respecting strided layouts
    for (size_t i = 0; i < m; ++i) {
        for (size_t j = 0; j < n; ++j) {
            T dot_sum = 0;
            for (size_t k = 0; k < k1; ++k) {
                dot_sum += (*this)({i, k}) * other({k, j});
            }
            out_data[i * n + j] = dot_sum;
        }
    }
    // 5. Returns resulting M x N ndarray
    return ndarray<T>(out_data, {m, n});
}

// Explicit template instantiations
template struct Storage<float>;
template struct Storage<double>;
template struct Storage<int>;

template class ndarray<float>;
template class ndarray<double>;
template class ndarray<int>;
